File: 518-coin-change-ii/coin-change-ii.cpp
class Solution {
public:
    int recur(vector<int> &coins, int amount, int i, vector<vector<int>> &dp){
        if(i < 0 || amount < 0) return 0;
        if(amount == 0) return 1;
        if(i == 0){
            if(amount % coins[i] == 0) return 1;
            return 0;
        }
        if(dp[i][amount] != -1) return dp[i][amount];
        return dp[i][amount] = recur(coins, amount - coins[i], i, dp) + recur(coins, amount, i-1, dp);
    }
    int change(int amount, vector<int>& coins) {
        vector<vector<int>> dp(coins.size(), vector<int>(amount+1, -1));
        return recur(coins, amount, coins.size()-1, dp);
    }
};

File: 518-coin-change-ii/coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:

        coins.sort()
        dp = [[-1] * (amount+1) for _ in range(len(coins)+1)]

        def recur(i, amt):
            if i < 0:
                return 0
            
            if amt == 0:
                return 1
            
            if dp[i][amt] != -1:
                return dp[i][amt]
            
            dp[i][amt] = 0
            if amt >= coins[i]:
                dp[i][amt] += recur(i, amt - coins[i])

            dp[i][amt] += recur(i-1, amt)

            return dp[i][amt]
        
        return recur(len(coins)-1, amount)
                    